src/python/pants/option/global_options.py

```diff
@@ -116,12 +145,39 @@ def from_bootstrap_options(cls, bootstrap_options: OptionValueContainer) -> Exec
             token_header = {"authorization": f"Bearer {oauth_token}"}
             remote_execution_headers.update(token_header)
             remote_store_headers.update(token_header)
+        if bootstrap_options.remote_auth_plugin and (
+            remote_execution or remote_cache_read or remote_cache_write
+        ):
+            auth_plugin_path, auth_plugin_func = bootstrap_options.remote_auth_plugin.split(":")
+            auth_plugin_module = importlib.import_module(auth_plugin_path)
+            auth_plugin_func = getattr(auth_plugin_module, auth_plugin_func)
+            auth_plugin_result = cast(
+                AuthPluginResult,
+                auth_plugin_func(
+                    initial_execution_headers=remote_execution_headers,
+                    initial_store_headers=remote_store_headers,
+                    options=options,
+                ),
+            )
+            # NB: We replace the headers variables so that the plugin author has total control
+            # over how the headers are set. We expect most plugins to preserve the original headers.
```

Comment:
Please document this semantic on the definitions of `execution_headers` and `store_headers` in `AuthPluginResult` as well.